v2/pkg/engine/datasource/pubsub_datasource/nats_event_manager.go

```diff
@@ -34,38 +40,75 @@ type NatsEventManager struct {
-	variablePlaceHolder, _ := p.variables.AddVariable(contextVariable) // $$0$$
-	return eventSubjectRegex.ReplaceAllLiteralString(subject, variablePlaceHolder), nil
+	variablePlaceHolder, _ := p.variables.AddVariable(contextVariable)
+	return variablePlaceHolder, nil
+}
+
+func (p *NatsEventManager) extractEventSubject(fieldRef int, subject string) (string, error) {
+	matches := argument_templates.ArgumentTemplateRegex.FindAllStringSubmatch(subject, -1)
+	// If no argument templates are defined, there are only static values
+	if len(matches) < 1 {
+		if natsserver.IsValidSubject(subject) {
+			return subject, nil
+		}
+		return "", fmt.Errorf(`subject "%s" is not a valid NATS subject`, subject)
+	}
+	fieldNameBytes := p.visitor.Operation.FieldNameBytes(fieldRef)
+	// TODO: handling for interfaces and unions
+	fieldDefinitionRef, ok := p.visitor.Definition.ObjectTypeDefinitionFieldWithName(p.visitor.Walker.EnclosingTypeDefinition.Ref, fieldNameBytes)
+	if !ok {
+		return "", fmt.Errorf(`expected field definition to exist for field "%s"`, fieldNameBytes)
+	}
+	subjectWithVariableTemplateReplacements := subject
+	for templateNumber, groups := range matches {
+		// The first group is the whole template; the second is the period delimited argument path
+		if len(groups) != 2 {
+			return "", fmt.Errorf(`argument template #%d defined on field "%s" is invalid: expected 2 matching groups but received %d`, templateNumber+1, fieldNameBytes, len(groups)-1)
+		}
+		validationResult, err := argument_templates.ValidateArgumentPath(p.visitor.Definition, groups[1], fieldDefinitionRef)
+		if err != nil {
+			return "", fmt.Errorf(`argument template #%d defined on field "%s" is invalid: %w`, templateNumber+1, fieldNameBytes, err)
+		}
+		argumentNameBytes := []byte(validationResult.ArgumentPath[0])
+		argumentRef, ok := p.visitor.Operation.FieldArgument(fieldRef, argumentNameBytes)
+		if !ok {
+			return "", fmt.Errorf(`operation field "%s" does not define argument "%s"`, fieldNameBytes, argumentNameBytes)
+		}
+		// variablePlaceholder has the form $$0$$, $$1$$, etc.
+		variablePlaceholder, err := p.addContextVariableByArgumentRef(
+			argumentRef, validationResult.ArgumentPath, validationResult.FinalInputValueTypeRef,
+		)
+		if err != nil {
+			return "", fmt.Errorf(`failed to retrieve variable placeholder for argument ""%s" defined on operation field "%s": %w`, argumentNameBytes, fieldNameBytes, err)
+		}
+		// Replace the template literal with the variable placeholder (and reuse the variable if it already exists)
+		subjectWithVariableTemplateReplacements = strings.ReplaceAll(subjectWithVariableTemplateReplacements, groups[0], variablePlaceholder)
+	}
+	// Substitute the variable templates for dummy values to check naïvely that the string is a valid NATS subject
+	if natsserver.IsValidSubject(variableTemplateRegex.ReplaceAllLiteralString(subjectWithVariableTemplateReplacements, "a")) {
```

Comment:
Is it clever here to create a dependency to natsserver? I'd copy the implementation of the function. This also reduce the binary size and compilation time. The server package is big.